scenarios/ebb_and_flow.go

```diff
@@ -0,0 +1,251 @@
+		sleepActivityConfig.Groups = map[string]loadgen.SleepActivityGroupConfig{"default": {}}
+	}
+	for name, group := range sleepActivityConfig.Groups {
+		fixedDist := loadgen.NewFixedDistribution(sleepDuration)
+		group.SleepDuration = &fixedDist
+		sleepActivityConfig.Groups[name] = group
+	}
+
+	var consecutiveErrCount int
+	errCh := make(chan error, 10000)
+	ticker := time.NewTicker(controlInterval)
+	defer ticker.Stop()
+	var startWG sync.WaitGroup
+	iter := 1
+
+	e.Logger.Infof("Starting ebb and flow scenario: min_backlog=%d, max_backlog=%d, frequency=%v, duration=%v",
+		minBacklog, maxBacklog, frequency, e.Configuration.Duration)
+
+	isDraining := false // true = draining mode, false = growing mode
+	cycleStartTime := e.startTime
+
+	for elapsed := time.Duration(0); elapsed < e.Configuration.Duration; elapsed = time.Since(e.startTime) {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err := <-errCh:
+			if err != nil {
+				e.Logger.Errorf("Failed to spawn workflow: %v", err)
+				consecutiveErrCount++
+				if consecutiveErrCount >= maxConsecutiveErrors {
+					return fmt.Errorf("got %v consecutive errors, most recent: %w", maxConsecutiveErrors, err)
+				}
+			} else {
+				consecutiveErrCount = 0
+			}
+		case <-ticker.C:
+			generated := e.generatedCount.Load()
+			processed := e.processedCount.Load()
+			backlog := generated - processed
+
+			// Check if we need to switch modes.
+			if isDraining && backlog <= minBacklog {
+				e.Logger.Infof("Backlog reached %d, switching to growing mode", backlog)
+				isDraining = false
+				cycleStartTime = time.Now()
+			} else if !isDraining && backlog >= maxBacklog {
+				e.Logger.Infof("Backlog reached %d, switching to draining mode", backlog)
+				isDraining = true
+				cycleStartTime = time.Now()
+			}
+
+			target := calculateBacklogTarget(isDraining, cycleStartTime, frequency, minBacklog, maxBacklog)
+			rate := calculateSpawnRate(target, backlog, minBacklog, maxBacklog, maxRate)
+
+			e.Logger.Infof("Backlog: %d, target: %d, rate: %d/s, gen: %d, proc: %d",
+				backlog, target, rate, generated, processed)
+
+			if rate > 0 {
+				startWG.Add(1)
+				go func(iteration, count int) {
+					defer startWG.Done()
+					errCh <- e.spawnWorkflowWithActivities(ctx, iteration, count, sleepActivityConfig)
+				}(iter, rate)
+				iter++
+			}
+		}
+	}
+
+	e.Logger.Info("Scenario complete; waiting for all workflows to finish...")
+	startWG.Wait()
+	return nil
+}
+
+func (e *ebbAndFlow) spawnWorkflowWithActivities(
+	ctx context.Context,
+	iteration, rate int,
+	template *loadgen.SleepActivityConfig,
+) error {
+	// Override activity count to fixed rate.
+	fixedDist := loadgen.NewFixedDistribution(int64(rate))
+	config := loadgen.SleepActivityConfig{
+		Count:  &fixedDist,
+		Groups: template.Groups,
+	}
+
+	// Generate activities.
+	var activities []*kitchensink.ExecuteActivityAction
+	activities = append(activities, config.Sample(e.rng)...)
+
+	// Start one workflow per activity.
```

Comment:
is this okay? is there any reason to try to batch multiple activities in a workflow?

Reply:
Initially, I was starting _all_ activities in one workflow. But I found it difficult to get the backlog target right since you would get a huge batch of activities finishing all at once. So I went to the most granular option. We could do _some_ batching, but now it becomes another parameter to configure and tweak.

Is your concern overloading the system with workflows and taken capacity away from the activities?